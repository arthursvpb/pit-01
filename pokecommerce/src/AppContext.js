import { createContext } from "react";
import { parseJwt, tokenKey } from "./utils/util";

const AppContext = createContext();

const searchParams = new URLSearchParams(window.location.search);

const initialState = {
  pokemons: [],
  wishlist: [],
  cart: [],
  pagination: {
    currentPage: searchParams.get("page") ?? 1,
    search: '',
    limit: 9,
    total: 0,
    totalPages: 0,
  },
  loggedUser: parseJwt(localStorage.getItem(tokenKey)),
};

const reducer = (state, action) => {
  switch (action.type) {
    case "SET_LOGGED_USER": {
      return {
        ...state,
        loggedUser: action.payload,
      };
    }

    case "SET_POKEMON": {
      return {
        ...state,
        pokemons: action.payload,
      };
    }

    case "SET_WISHLIST": {
      return {
        ...state,
        wishlist: action.payload,
      };
    }

    case "SET_PAGINATION": {
      return {
        ...state,
        pagination: action.payload,
      };
    }

    default: {
      return state;
    }
  }
};

export { initialState, reducer };

export default AppContext;
