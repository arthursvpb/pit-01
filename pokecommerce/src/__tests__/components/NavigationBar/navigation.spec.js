import NavigationBar from "../../../components/NavigationBar";

import AppContext, { initialState } from "../../../AppContext";

import { render, fireEvent } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

const NavigationBarWrapper = ({ state = initialState, dispatch }) => (
  <AppContext.Provider value={[state, dispatch]}>
    <BrowserRouter>
      <NavigationBar />
    </BrowserRouter>
  </AppContext.Provider>
);

describe("NavigationBar", () => {
  it("renders", () => {
    const { asFragment } = render(<NavigationBarWrapper />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with context data", () => {
    const dispatch = jest.fn();

    const { container, queryByText } = render(
      <NavigationBarWrapper
        state={{
          ...initialState,
          loggedUser: {
            name: "Keven Leone",
          },
          me: {
            pokeDolar: 1000,
          },
        }}
        dispatch={dispatch}
      />
    );

    expect(container.querySelectorAll(".dynamic-route")).toHaveLength(3);
    expect(queryByText("Pokedolar $1000")).toBeTruthy();
    expect(queryByText("Welcome: Keven Leone")).toBeTruthy();

    const logoutButton = queryByText("Logout");

    fireEvent.click(logoutButton);

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenLastCalledWith({
      type: "SET_LOGGED_USER",
      payload: null,
    });
  });
});
