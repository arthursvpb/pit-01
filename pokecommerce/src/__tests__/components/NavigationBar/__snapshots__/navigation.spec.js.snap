// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBar renders 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-collapse-absolute navbar-expand-md navbar-underline navigation-bar navigation-bar-light"
  >
    <div
      class="container-fluid container-fluid-max-xl"
    >
      <button
        aria-expanded="false"
        class="navbar-toggler navbar-toggler-link collapsed btn btn-unstyled"
        type="button"
      >
        <span
          class="navbar-text-truncate"
        >
          My Pokemon Commerce
        </span>
        <svg
          class="lexicon-icon lexicon-icon-caret-bottom"
          role="presentation"
        >
          <use
            xlink:href="#caret-bottom"
          />
        </svg>
      </button>
      <div
        class="navbar-collapse collapse"
      >
        <div
          class="container-fluid container-fluid-max-xl"
        >
          <ul
            class="navbar-nav"
          >
            <li
              class="nav-item dynamic-route"
            >
              <a
                class="nav-link active link-unstyled"
              >
                Pokemons
              </a>
            </li>
            <li
              class="nav-item dynamic-route"
            >
              <a
                class="nav-link link-unstyled"
              >
                Wishlist
              </a>
            </li>
            <li
              class="nav-item dynamic-route"
            >
              <a
                class="nav-link link-unstyled"
              >
                Purchased Pokemon
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link link-unstyled"
              >
                Pokedolar $
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
